File: flint/execution/synthetic_depth.py
"""Synthetic orderbook depth model.

Generates realistic orderbook snapshots from per-venue depth profiles.
Fallback when real data (Drift S3, Hyperliquid archive, Tardis) is unavailable.
"""
from __future__ import annotations
import math
from dataclasses import dataclass
from typing import Dict
from ..models import OrderbookLevel, OrderbookSnapshot


@dataclass(frozen=True)
class DepthProfile:
    """Venue depth characteristics for BTC-PERP baseline.
    Other markets scale linearly by relative volume.
    """
    bid_depth_1pct: float    # USD liquidity within 1% of mid (bid side)
    ask_depth_1pct: float    # USD liquidity within 1% of mid (ask side)
    concentration: float     # How concentrated at top-of-book (0-1)
    spread_bps: float        # Typical bid-ask spread in basis points


VENUE_PROFILES: Dict[str, DepthProfile] = {
    "binance": DepthProfile(20_000_000, 20_000_000, 0.7, 0.5),
    "coinbase": DepthProfile(15_000_000, 15_000_000, 0.65, 0.8),
    "okx": DepthProfile(10_000_000, 10_000_000, 0.6, 1.0),
    "bybit": DepthProfile(8_000_000, 8_000_000, 0.6, 1.0),
    "kraken": DepthProfile(4_000_000, 4_000_000, 0.55, 1.5),
    "kucoin": DepthProfile(6_000_000, 6_000_000, 0.55, 1.2),
    "bitget": DepthProfile(6_000_000, 6_000_000, 0.55, 1.2),
    "gate": DepthProfile(3_000_000, 3_000_000, 0.5, 2.0),
    "mexc": DepthProfile(2_000_000, 2_000_000, 0.45, 2.5),
    "htx": DepthProfile(3_000_000, 3_000_000, 0.5, 2.0),
    "hyperliquid": DepthProfile(5_000_000, 5_000_000, 0.5, 1.5),
    "drift": DepthProfile(2_000_000, 2_000_000, 0.4, 3.0),
}


def generate_synthetic_book(
    mid_price: float, profile: DepthProfile, levels: int = 20,
    market: str = "", ts: int = 0,
) -> OrderbookSnapshot:
    """Generate a synthetic orderbook from a depth profile.

    Uses exponential decay controlled by concentration parameter.
    """
    half_spread = (profile.spread_bps / 10_000) * mid_price / 2
    step = (mid_price * 0.01) / levels  # 1% range / N levels

    decay = 1.0 + profile.concentration * 4  # 0→1 maps to 1→5
    weights = [math.exp(-decay * i / levels) for i in range(levels)]
    total_weight = sum(weights)

    bids, asks = [], []
    for i in range(levels):
        frac = weights[i] / total_weight
        bid_price = mid_price - half_spread - i * step
        bid_usd = profile.bid_depth_1pct * frac
        bid_size = bid_usd / bid_price if bid_price > 0 else 0
        bids.append(OrderbookLevel(price=round(bid_price, 6), size=round(bid_size, 4)))

        ask_price = mid_price + half_spread + i * step
        ask_usd = profile.ask_depth_1pct * frac
        ask_size = ask_usd / ask_price if ask_price > 0 else 0
        asks.append(OrderbookLevel(price=round(ask_price, 6), size=round(ask_size, 4)))

    return OrderbookSnapshot(market=market, ts=ts, bids=tuple(bids), asks=tuple(asks))

File: rust/src/engine/synthetic_depth.rs
//! Synthetic orderbook generation from per-venue depth profiles.

use crate::types::BookLevel;

#[derive(Debug, Clone, Copy)]
pub struct DepthProfile {
    pub bid_depth_1pct: f64,
    pub ask_depth_1pct: f64,
    pub concentration: f64,
    pub spread_bps: f64,
}

/// Lookup depth profile by venue name.
pub fn get_depth_profile(venue: &str) -> DepthProfile {
    match venue {
        "binance"     => DepthProfile { bid_depth_1pct: 20_000_000.0, ask_depth_1pct: 20_000_000.0, concentration: 0.7, spread_bps: 0.5 },
        "coinbase"    => DepthProfile { bid_depth_1pct: 15_000_000.0, ask_depth_1pct: 15_000_000.0, concentration: 0.65, spread_bps: 0.8 },
        "okx"         => DepthProfile { bid_depth_1pct: 10_000_000.0, ask_depth_1pct: 10_000_000.0, concentration: 0.6, spread_bps: 1.0 },
        "bybit"       => DepthProfile { bid_depth_1pct: 8_000_000.0, ask_depth_1pct: 8_000_000.0, concentration: 0.6, spread_bps: 1.0 },
        "kraken"      => DepthProfile { bid_depth_1pct: 4_000_000.0, ask_depth_1pct: 4_000_000.0, concentration: 0.55, spread_bps: 1.5 },
        "kucoin"      => DepthProfile { bid_depth_1pct: 6_000_000.0, ask_depth_1pct: 6_000_000.0, concentration: 0.55, spread_bps: 1.2 },
        "bitget"      => DepthProfile { bid_depth_1pct: 6_000_000.0, ask_depth_1pct: 6_000_000.0, concentration: 0.55, spread_bps: 1.2 },
        "gate"        => DepthProfile { bid_depth_1pct: 3_000_000.0, ask_depth_1pct: 3_000_000.0, concentration: 0.5, spread_bps: 2.0 },
        "mexc"        => DepthProfile { bid_depth_1pct: 2_000_000.0, ask_depth_1pct: 2_000_000.0, concentration: 0.45, spread_bps: 2.5 },
        "htx"         => DepthProfile { bid_depth_1pct: 3_000_000.0, ask_depth_1pct: 3_000_000.0, concentration: 0.5, spread_bps: 2.0 },
        "hyperliquid" => DepthProfile { bid_depth_1pct: 5_000_000.0, ask_depth_1pct: 5_000_000.0, concentration: 0.5, spread_bps: 1.5 },
        "drift"       => DepthProfile { bid_depth_1pct: 2_000_000.0, ask_depth_1pct: 2_000_000.0, concentration: 0.4, spread_bps: 3.0 },
        _             => DepthProfile { bid_depth_1pct: 5_000_000.0, ask_depth_1pct: 5_000_000.0, concentration: 0.5, spread_bps: 1.5 },
    }
}

/// Generate synthetic orderbook from a depth profile.
pub fn generate_synthetic_book(
    mid_price: f64,
    profile: &DepthProfile,
    levels: usize,
) -> (Vec<BookLevel>, Vec<BookLevel>) {
    let half_spread = (profile.spread_bps / 10_000.0) * mid_price / 2.0;
    let step = (mid_price * 0.01) / levels as f64;

    let decay = 1.0 + profile.concentration * 4.0;
    let weights: Vec<f64> = (0..levels)
        .map(|i| (-decay * i as f64 / levels as f64).exp())
        .collect();
    let total_weight: f64 = weights.iter().sum();

    let mut bids = Vec::with_capacity(levels);
    let mut asks = Vec::with_capacity(levels);

    for i in 0..levels {
        let frac = weights[i] / total_weight;

        let bid_price = mid_price - half_spread - i as f64 * step;
        let bid_usd = profile.bid_depth_1pct * frac;
        let bid_size = if bid_price > 0.0 { bid_usd / bid_price } else { 0.0 };
        bids.push(BookLevel { price: bid_price, size: bid_size });

        let ask_price = mid_price + half_spread + i as f64 * step;
        let ask_usd = profile.ask_depth_1pct * frac;
        let ask_size = if ask_price > 0.0 { ask_usd / ask_price } else { 0.0 };
        asks.push(BookLevel { price: ask_price, size: ask_size });
    }

    (bids, asks)
}

/// Scale a depth profile by bar volume ratio.
pub fn scale_profile(profile: &DepthProfile, bar_volume_usd: f64) -> DepthProfile {
    const VOL_DEPTH_RATIO: f64 = 20.0;
    const FLOOR: f64 = 0.05;
    const CAP: f64 = 3.0;

    if bar_volume_usd <= 0.0 {
        return *profile;
    }

    let expected_vol = profile.bid_depth_1pct * VOL_DEPTH_RATIO;
    let scale = (bar_volume_usd / expected_vol).clamp(FLOOR, CAP);

    DepthProfile {
        bid_depth_1pct: profile.bid_depth_1pct * scale,
        ask_depth_1pct: profile.ask_depth_1pct * scale,
        ..*profile
    }
}
